lib/plugin.c

```diff
@@ -168,7 +174,235 @@ static struct mptcpd_plugin_ops const *token_to_ops(mptcpd_token_t token)
 //                Plugin Registration and Management
 // ----------------------------------------------------------------
 
-bool mptcpd_plugin_load(char const *dir, char const *default_name)
+/**
+ * @struct plugin_info
+ *
+ * @brief Plugin information
+ */
+struct plugin_info
+{
+        /// Handle returned from call to @c dlopen().
+        void *handle;
+
+        /// Plugin descriptor.
+        struct mptcpd_plugin_desc const *desc;
+};
+
+/// List of @c plugin_info objects.
+static struct l_queue *_plugin_infos;
+
+/**
+ * @brief Compare plugin priorities.
+ *
+ * @param[in] a         New      plugin information.
+ * @param[in] b         Existing plugin information.
+ * @param[in] user_data User data (unused).
+ *
+ * @return Value greater than 0 if the new plugin should be inserted
+ *         after the existing plugin in the list, and a negative value
+ *         otherwise.
+ */
+static int compare_plugin_priority(void const *a,
+                                   void const *b,
+                                   void *user_data)
+{
+        (void) user_data;
+
+        struct plugin_info const *const new      = a;
+        struct plugin_info const *const existing = b;
+
+        /*
+          Cause "new" plugin to be inserted into the plugin list
+          before the "existing" plugin it its priority is higher
+          (lower value) than the existing plugin priority.
+        */
+        return new->desc->priority - existing->desc->priority;
```

Comment:
Does it make sense to enforce a valid range of priorities? Using subtraction here allows some odd integer wraparound issues if someone uses values near INT_MAX or INT_MIN. Or could allow any ints and use a comparison operator instead.

Reply:
That's a good point.  I'll use a comparison operator instead.  Consequently, should we make the low and high priority values be `INT_MIN` and `INT_MAX`, respectively?

Reply:
I think the existing low/high values in the ~20 range are fine and encourage people to use a similar range of values. But allowing the full INT range still lets them use whatever value they want.

Reply:
Commit 7731e6c415ae9df2339bc359784ceb8d81eaabd9 allows the full `int` range to be used, while retaining the small default low/high priorities around  +/-20.